src/Misc/coroutine.ts

```diff
@@ -0,0 +1,175 @@
+        if (!abortSignal || !abortSignal.aborted) {
+            scheduler(coroutine,
+                (stepResult: CoroutineStep<T>) => {
+                    if (stepResult.done) {
+                        // If the coroutine is done, report success.
+                        onSuccess(stepResult.value);
+                    } else {
+                        // If the coroutine is not done, resume the coroutine (via the scheduler).
+                        resume();
+                    }
+                },
+                (error: any) => {
+                    // If the coroutine threw an error, report the error.
+                    onError(error);
+                });
+        } else {
+            onError("Aborted");
+        }
+    }
+
+    resume();
+}
+
+// Runs the specified coroutine synchronously.
+/** @hidden */
+export function runCoroutineSync<T>(coroutine: Coroutine<T>, abortSignal?: AbortSignal): T {
+    // Run the coroutine with the inline scheduler, storing the returned value, or re-throwing the error (since the error callback will be called synchronously by the inline scheduler).
+    let result: T | undefined;
+    runCoroutine(coroutine, inlineScheduler, (r: T) => result = r, (e: any) => { throw e; }, abortSignal);
+
+    // Synchronously return the result of the coroutine.
+    return result!;
+}
+
+// Runs the specified coroutine asynchronously with the specified scheduler.
+/** @hidden */
+export function runCoroutineAsync<T>(coroutine: AsyncCoroutine<T>, scheduler: CoroutineScheduler<T>, abortSignal?: AbortSignal): Promise<T> {
+    // Run the coroutine with a yielding scheduler, resolving or rejecting the result promise when the coroutine finishes.
+    return new Promise((resolve, reject) => {
+        runCoroutine(coroutine, scheduler, resolve, reject, abortSignal);
+    });
+}
+
+// This is a helper type to extract the return type of a Coroutine<T>. It is conceptually very similar to the Awaited<T> utility type.
+/** @hidden */
+type ExtractCoroutineReturnType<T> =
+    T extends Coroutine<infer TReturn> ? TReturn :
+    never;
+
+// This is a helper type to extract the return type of an AsyncCoroutine<T>.
+/** @hidden */
+type ExtractAsyncCoroutineReturnType<T> =
+    T extends Coroutine<infer TReturn> ? Promise<TReturn> :
+    T extends AsyncCoroutine<infer TReturn> ? Promise<TReturn> :
+    never;
+
+/**
+ * Given a function that returns a Coroutine<T>, produce a function with the same parameters that returns a T.
+ * The returned function runs the coroutine synchronously.
+ * @param coroutineFactory A function that returns a Coroutine<T>.
+ * @returns A function that runs the coroutine synchronously.
+ * @hidden
+ */
+export function makeSyncFunction<TReturn, TCoroutineFactory extends (...params: any[]) => Coroutine<TReturn>>(coroutineFactory: TCoroutineFactory, abortSignal?: AbortSignal): (...params: Parameters<TCoroutineFactory>) => ExtractCoroutineReturnType<ReturnType<TCoroutineFactory>> {
```

Comment:
It is a little weird to me that these are just sort of out in the global namespace. When imported under an alias (as is done in the Playground, for example), how will these show up? Will they just be `BABYLON.makeSyncFunction(...)`?

Reply:
Namespacing or making these static functions on a type or something might also help clarify the connection between these and coroutines, I think.

Reply:
This is a good point. What happens to these functions in the documentation?

Reply:
@RaananW it sounds like a good topic for you ?

Reply:
You are exactly right that these will show up as `BABYLON.makeSyncFunction(...)` for the UMD package (see the unit tests). That said, my understanding is that this is the direction BJS is actively moving to improve tree shaking (since otherwise if you just need one function in a class containing a bunch of static functions, all those functions get brought in). For ES6 modules, these functions are naturally "name spaced" by virtue of being in a module. I'm not sure if there is a way to group these functions beyond modules without hurting tree shaking. @RaananW can you confirm? Also, is there a naming scheme planned to make this more usable in UMD? For example, if we had something like `Mesh.Merge` and it was moved out to just be an exported function, would it just be named `Merge`, or is there some naming scheme we want to make it sort of name spaced within the context of UMD?

Reply:
Yes, the UMD package will contain those as a "global" member, because of the way we build it (anything exported is automatically a member of the BABYLON namespace).

What we can consider for the UMD LTS version is to remove the global declaration of functions assigned to static members, so that they at least don't appear in the typedoc of the BABYLON namespace. js-wise they will still be there though. I won't change the way we build. I don't see that as something that is needed for a few reasons. The main one is that UMD should slowly be a "legacy" package that is only used in very few cases, and the second is that almost any other framework (with UMD/Namespacing support" does the same - all functions are part of the global namespace.

Anyhow - we should not use static functions (especially not private static, which makes no sense when working with modules), but this is a topic I will first address in January (during the build system overhaul). They should be out of the class context to support better tree shaking.

Regarding naming - this is still an open question that I need to address soon. As the modules are still constructed based on directory, this will still be possible:

```javascript
import { Merge } from "@packageName/Meshes/mesh";
```
Instead of loading the Mesh class, you are loading from the Mesh module.

However, this would also be possible (while still supporting tree shaking, because there are no more side effects):

```javascript
import { Merge } from "@packageName";
```

This depends on whether or not we also provide an index file, but that's an internal decision I still need to make after experimenting with building the new esm package. This will eventually influence naming, as we can't have two functions named Merge (or - a better example is the `Parse` function that a lot of different components have)

TL;dr - I will run a pass at the static structure once I have all of these answers. For now - if it is assigned to a member of a class don't export it from the module and keep the class reference. This will take it out of the exported global namespace. If it is a new function export it and use it directly from the global namespace.